### DISCUSSION TITLE
Why can't out and params be used together?

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#5273] [General]
I have a situation where I'm generating C# code from SQL stored procedure metadata.

Id like to be able to define a method as:

` Invoke(out params (string Name, object Value)[] Arglist)`

But this is not supported (removing the 'out' of course makes it fine).

Is there a deep reason? is this something that could be added to the language?

The root of this is that SQL stored procedures can themselves have 'out' parameters.

**Comment by CyrusNajmabadi** [comment]
What would this mean?  How would it work?  If you wrote code that called this, what would it compile down to?

**Reply by jnm2** [reply]
Maybe he wants the ability for the caller to do `M(out var a, out var b, ...)` to any arity without having to define an overload per arity?

**Reply by The-Futurist** [reply]
> Maybe he wants the ability for the caller to do `M(out var a, out var b, ...)` to any arity without having to define an overload per arity?

Exactly, that was the ultimate motive. Here's an example of a generated method:

```cs
        public List<dt_getpropertiesbyid_vcs_u> Invoke(int @id, string @property, out string @value)
        {
            @value = default;
            var arguments = new List<Argument>
            {
                new Argument(nameof(@id),@id,false),
                new Argument(nameof(@property),@property,false),
                new Argument(nameof(@value),@value,true),
            };
            var results = base.Invoke(arguments);
            @value = (string)arguments.Where(a => a.Name == "@value").First().Value;
            return results;
        }
```

This and every other generated method calls the base:

```cs
        protected List<T> Invoke(List<Argument> Arguments)
```
I became a little fixated on the "out" as I was thinking about all that, of course there was no need, nor a need for tuples.

(Actually these signatures really should be using `ref` not `out` because I discovered yesterday that an OUT in SQL Server actually always _really_ means IN/OUT).

**Comment by DavidArno** [comment]
What are you trying to achieve with `params` there? [If you remove it, it'll work the way I think you want it to work](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBfgCwAEAkjAG4D2A1gKYAULqCtXfMQAMtAHIQwHOLRbAAVhwDG/AGoQANqg4BKANoBdWgEFkAcw1QAzgh0UA3hVrOT5yzdoBeWjA4B3ASFhGTlFFX0DJxdHchc4gQAiAHlfBJliHTgo+MSAFT8WNNpCO1iXAF8AbgpyigpGVk4ePlomCGRaZA4rVA0EKx1q8gAzFi6IJQALATaOrhhJaVbNbR1aWE7u3v7SmLihAE4uABIE+wWpcq9aezatDnKEwZqgA):

```c#
void Invoke(out (string Name, object Value)[] Arglist)
{
    Arglist = new (string, object)[]
    {
        ("One", 1),
        ("Two", 2)
    };
}

Invoke(out var results);
foreach (var (name, value) in results)
{
    Console.WriteLine($"{name} = {value}");
}
```

**Comment by The-Futurist** [comment]
@DavidArno - I was a little hasty, ultimately I am building a code generator and I wanted to be able to call into a common base method that is wholly general purpose.

Basically we might have generated this:

```cs
public IEnumerable<T> Invoke (int Arg1, double Arg2, out DateTime Arg3) 
{ 
// code... 
}
```
and this
```cs
public IEnumerable<T> Invoke (string Arg1, out double Arg2, out long Arg3)
{ 
// code... 
}
```

and in each case, in each of these generated method bodies, make a call to a more general method - so I was looking for a method in which every single arg was an 'out' arg and I could then support any signature.

Of course you are correct, there's absolutely no need to support what I was thinking of, values can be passed back without problems and there's no need for "out" here at all.

**Comment by The-Futurist** [comment]
Incidentally, I stumbled upon this unexpected behavior yesterday:

```cs
using System;
using System.Collections.Generic;

namespace ConsoleApp3
{
    class Program
    {
        static void Main(string[] args)
        {
            var objects = new List<Argument>() 
            { 
                new ("A", 123, true),
                new ("B", 234, false),
                new ("C", 456, true),
                new ("D", 678, false)
            };

            var tuples = new List<(string Name, object Value, bool Out)>()
            {
                ("A", 123, true),
                ("B", 234, false),
                ("C", 456, true),
                ("D", 678, false)
            };

            foreach (var o in objects)
            {
                if (o.Name == "B")
                    o.Value = 100;
            }

            foreach (var t in tuples)
            {
                if (t.Name == "B")
                    t.Value = 100;
            }
        }
    }

    public class Argument
    {
        public Argument(string Name, object Value, bool Out)
        {
            this.Name = Name;
            this.Value = Value;
            this.Out = Out;
        }
        public string Name { get; private set; }
        public object Value { get; internal set; }
        public bool Out { get; private set; }
    }
}
```

The assignment to `Value` inside the second loop generates [CS1654](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1654).

**Reply by jnm2** [reply]
That's because ValueTuple is a struct. `t` has already been copied and is not a reference back into the list. You'd see the same thing if you changed Argument to a struct.
